src/test/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounderTest.java

```diff
@@ -97,14 +112,228 @@ public void groundConstraintAlreadyGround() {
 		Program program = PARSER.parse("a :- not b. "
 				+ "b :- not a. "
 				+ ":- b.");
-		
+
 		AtomStore atomStore = new AtomStoreImpl();
 		Grounder grounder = GrounderFactory.getInstance("naive", program, atomStore, true);
 		Map<Integer, NoGood> noGoods = grounder.getNoGoods(new TrailAssignment(atomStore));
 		int litB = Literals.atomToLiteral(atomStore.get(new BasicAtom(Predicate.getInstance("b", 0))));
-		assertTrue(noGoods.containsValue(NoGood.fromConstraint(Arrays.asList(litB), Collections.emptyList())));
+		assertTrue(noGoods.containsValue(NoGood.fromConstraint(Collections.singletonList(litB), Collections.emptyList())));
+	}
+
+	@Test
+	public void avoidDeadEndsWithLaxGrounderHeuristic() {
+		RuleGroundingOrder groundingOrderP1 = new RuleGroundingOrder(literal("p1", "X"),
+				Arrays.asList(literal("p2", "X"), literal("q2", "Y"), literal("q1", "Y")), -1, false);
+		RuleGroundingOrder groundingOrderQ1 = new RuleGroundingOrder(literal("q1", "Y"),
+				Arrays.asList(literal("q2", "Y"), literal("p2", "X"), literal("p1", "X")), -1, false);
+		testDeadEnd(groundingOrderP1, groundingOrderQ1, true);
+	}
+
+	@Test
+	public void noDeadEndWithLaxGrounderHeuristic() {
+		RuleGroundingOrder groundingOrderP1 = new RuleGroundingOrder(literal("p1", "X"),
+				Arrays.asList(literal("p2", "X"), literal("q1", "Y"), literal("q2", "Y")), -1, false);
+		RuleGroundingOrder groundingOrderQ1 = new RuleGroundingOrder(literal("q1", "Y"),
+				Arrays.asList(literal("q2", "Y"), literal("p1", "X"), literal("p2", "X")), -1, false);
+		testDeadEnd(groundingOrderP1, groundingOrderQ1, true);
+	}
+
+	private void testDeadEnd(RuleGroundingOrder groundingOrderP1, RuleGroundingOrder groundingOrderQ1, boolean expectNoGoods) {
```

Comment:
It would be good if those private helpers were a bit documented.

Reply:
Done.